### PR TITLE
Scroll bars now optional

### PR DESCRIPTION
**Pull request by JordanMartinez**
Ok. Let's try this again....

### CODE REVIEW

**Comment by JordanMartinez** [comment]
Since I asked about the VirtualizedScrollPane#dispose method in the previous problematic PR, I figured we should bring it here since most would look here not there.

I asked:

> Should the `VirtualizedScrollPane#dispose()` method be included?

You replied:
> Hmm, dispose() in the current VirtualFlow only calls content.dispose(), which stops observing the list of items. That kind of dispose doesn't make sense for VirtualizedScrollPane. However there is a number of bindings created that are not disposed—they don't need to be disposed, because currently VirtualFlow and VirtualFlowContent always come together. However, if we allow to remove the content from VirtualizedScrollPane, we should dispose those bindings. Preferably, however, they should be disposed automatically when content is removed.

My response:
Ok. So, should content be held as a `SimpleObjectProperty<V> content` with a listener for its removal? Or should there be the method `removeContent()` that unbinds from the content and then returns `V content` ?

**Comment by TomasMikula** [comment]
I think the latter is simpler for us to do. There's a catch for both options, though: [Node](http://docs.oracle.com/javase/8/javafx/api/javafx/scene/Node.html) documentation says:

> If a program adds a child node to a Parent (including Group, Region, etc) and that node is already a child of a different Parent or the root of a Scene, the node is automatically (and silently) removed from its former parent.

So the user will be able to remove the content from `VirtualizedScrollPane` even without calling `removeContent()`&mdash;by simply adding it to a different parent. We should be able to detect that as well and do the clean-up automatically.

**Comment by JordanMartinez** [comment]
True....
So, something like:
````java
content.parentProperty().addListener(changed -> dispose());

// where dispose is:
    private void dispose() {
        unbindScrollBar(hbar);
        unbindScrollBar(vbar);
    }
    
    private void unbindScrollBar(ScrollBar bar) {
        bar.maxProperty().unbind();
        bar.unitIncrementProperty().unbind();
        bar.blockIncrementProperty().unbind();
        bar.valueProperty().unbind();
    }
````

**Comment by TomasMikula** [comment]
Mmm, who removes the `content.parentProperty()`'s listener then? I was thinking more like

```java
ObjectBinding<V> content = (ObjectBinding<V>) Bindings.valueAt(getChildren(), 0);
content.addListener(changed -> dispose());
```

Here we don't need to call `content.dispose()` or `content.removeListener(...)`, since we're only observing our own fields.

**Comment by TomasMikula** [comment]
Or just observe the `getChildren()` list for changes. Once populated (with `content`), the only possible change is `content` being removed, so when a change occurs, dispose.

**Comment by TomasMikula** [comment]
`removeContent` would then be implemented as

```java
getChildren().clear();
return content;
```

**Comment by JordanMartinez** [comment]
What would the code for listening to the changes in children's size be? I understand that we want to avoid memory leaks by removing the listener once finished. However, `ListChangeListener` has always felt weird to me. Would you be against me using ReactFX's `EventStreams.sizesOf(getChildren()).subscribe( () -> dispose());` and then calling the resulting Subscription's `unsubscribe()` method in the `dispose()` method? Or is there a better way to do this?

**Comment by JordanMartinez** [comment]
Do these need to be unbound as well?
- [Lines 44-49](https://github.com/JordanMartinez/Flowless/blob/scrollBarsNowOptional/src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java#L44-L49)
- [Lines 54-75](https://github.com/JordanMartinez/Flowless/blob/scrollBarsNowOptional/src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java#L54-L75)

**Comment by TomasMikula** [comment]
I agree they are weird. An invalidation listener on `getChildren()` should suffice, though.

**Comment by TomasMikula** [comment]
- [Lines 44-49](https://github.com/JordanMartinez/Flowless/blob/scrollBarsNowOptional/src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java#L44-L49)

Yes.

- [Lines 54-75](https://github.com/JordanMartinez/Flowless/blob/scrollBarsNowOptional/src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java#L54-L75)

No, these will be unbound automatically as soon as [L76-L77](https://github.com/JordanMartinez/Flowless/blob/scrollBarsNowOptional/src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java#L76-L77) are unbound. That means L76-L77 need to be unbound.

**Comment by JordanMartinez** [comment]
Ok. Here's what I have:
````java
class VirtualizedScrollPane /* .... */ {
    private Var<Double> hbarValue;
    private Var<Double> vbarValue;

    VirtualizedScrollPane(V content) {
        // constructor stuff

        // scrollbar positions
        hbarValue = Var.doubleVar(hbar.valueProperty());
        vbarValue = Var.doubleVar(vbar.valueProperty());
        Bindings.bindBidirectional(
                hbarValue,
                content.estimatedScrollXProperty());
        Bindings.bindBidirectional(
                vbarValue,
                content.estimatedScrollYProperty());
        // and the rest that is normally there...

        getChildren().addListener(new InvalidationListener() {
            @Override
            public void invalidated(Observable observable) {
                dispose();
            }
        });
    }


    /**
     * Unbinds scrolling from Content before returning Content.
     * @return - the content
     */
    V removeContent() {
        getChildren().clear();
        return content;
    }

    private void dispose() {
        hbarValue.unbindBidirectional(content.estimatedScrollXProperty());
        vbarValue.unbindBidirectional(content.estimatedScrollYProperty());
        unbindScrollBar(hbar);
        unbindScrollBar(vbar);
    }

    private void unbindScrollBar(ScrollBar bar) {
        bar.maxProperty().unbind();
        bar.unitIncrementProperty().unbind();
        bar.blockIncrementProperty().unbind();
        bar.valueProperty().unbind();
    }
````

**Comment by TomasMikula** [comment]
This seems unnecessary:

    bar.valueProperty().unbind();

OTOH, you need to add

    bar.visibleProperty().unbind();

**Comment by TomasMikula** [comment]
Also please port Javadoc comments from the [original `VirtualFlow.java`](https://github.com/TomasMikula/Flowless/blob/6012497c54b7511f23dbd3e3a665ed32ed4af547/src/main/java/org/fxmisc/flowless/VirtualFlow.java) to the new one where applicable.

**Comment by JordanMartinez** [comment]
Oh, that was supposed to be `visibleProperty()`

Ok. I'll take care of that.

**Comment by TomasMikula** [comment]
It seems that [`cellToViewport` methods](https://github.com/TomasMikula/Flowless/blob/6012497c54b7511f23dbd3e3a665ed32ed4af547/src/main/java/org/fxmisc/flowless/VirtualFlow.java#L154-L164) also got lost in translation.

**Comment by TomasMikula** [comment]
A good sanity check would be to test RichTextFX with this new Flowless.

**Comment by JordanMartinez** [comment]
I'm checking it right now.

**Comment by JordanMartinez** [comment]
VirtualFlow isn't public so it can't be accessed outside of its package.... Fixing that.

**Comment by JordanMartinez** [comment]
Yeah... there's a lot of things that need to be made public now.

**Comment by JordanMartinez** [comment]
If StyledTextArea implements Virtualized, it conflicts with totalWidth/HeightEstimateProperty methods because they return `DoubleProperty` instead of `Val<Double>`

**Comment by JordanMartinez** [comment]
Otherwise, it seems to work fine. Once the skin gets removed, the actual properties from VirtualFlow could be returned.

**Comment by JordanMartinez** [comment]
I found a bug. If I copy a text using the RichText demo and paste it. I get this:
````java
Exception in thread "JavaFX Application Thread" java.lang.NullPointerException
	at org.fxmisc.richtext.demo.richtext.RichText.lambda$null$24(RichText.java:154)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:545)
	at java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
	at java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:438)
	at org.fxmisc.richtext.demo.richtext.RichText.lambda$start$29(RichText.java:154)
	at org.reactfx.value.ChangeListenerWrapper.accept(Val.java:758)
	at org.reactfx.util.AbstractReducingStreamNotifications.lambda$head$275(NotificationAccumulator.java:248)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
	at org.reactfx.value.ValBase.invalidate(ValBase.java:32)
	at org.reactfx.SuspendableBoolean.release(SuspendableBoolean.java:24)
	at org.reactfx.CloseableOnceGuard.close(Guard.java:49)
	at org.reactfx.MultiGuard.close(Guard.java:83)
	at org.fxmisc.richtext.StyledTextArea.replace(StyledTextArea.java:804)
	at org.fxmisc.richtext.TextEditingArea.replace(TextEditingArea.java:212)
	at org.fxmisc.richtext.EditActions.replaceSelection(EditActions.java:129)
	at org.fxmisc.richtext.ClipboardActions.paste(ClipboardActions.java:88)
	at org.fxmisc.richtext.skin.StyledTextAreaBehavior.lambda$static$70(StyledTextAreaBehavior.java:65)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$On.lambda$null$23(EventHandlerTemplate.java:156)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$On.lambda$act$24(EventHandlerTemplate.java:155)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$Builder.lambda$create$21(EventHandlerTemplate.java:117)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$17(EventHandlerTemplate.java:39)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$19(EventHandlerTemplate.java:49)
````

**Comment by JordanMartinez** [comment]
False alarm. That's a bug in the current version.

**Comment by TomasMikula** [comment]
Yeah, don't bother `StyledTextArea` implementing `Virtualized` before we get rid of skin. Whether or not the rendering is virtualized is now completely up to the skin.

Yeah, there is a bug. But AFAIK, the bug is not present in the released 0.6.10 version, only in the current master. This is due to half-baked incomplete TomasMikula/RichTextFX#6. I should have tested it better before merging.

**Comment by JordanMartinez** [comment]
I see you've fixed the copy/paste with paragraph styles.

**Comment by TomasMikula** [comment]
Yes. If you are happy with this PR and there aren't any troubles porting RichTextFX to this version of Flowless, I'm ready to merge this.

**Comment by JordanMartinez** [comment]
This might be a stupid question, but is this supposed to happen?
When I write the following code in StyledTextAreaView's constructor:
````java
VirtualFlow virtuaFlow = // its constructor...

VirtualizedScrollPane<VirtualFlow> vsPane = new VirtualizedScrollPane(virtualFlow);
getChildren().add(vsPane);
````

Any text is no longer displayed, even when area is initialized with text.

**Comment by TomasMikula** [comment]
Well, no, that's not supposed to happen :D

File: src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+                layoutHeight,
+                (needsV, needsH, ch, hbh, lh) -> needsV || needsH && ch + hbh.doubleValue() > lh);
+        hbar.visibleProperty().bind(needsHBar);
+        vbar.visibleProperty().bind(needsVBar);
+
+        // request layout later, because if currently in layout, the request is ignored
+        hbar.visibleProperty().addListener(obs -> Platform.runLater(() -> requestLayout()));
+        vbar.visibleProperty().addListener(obs -> Platform.runLater(() -> requestLayout()));
+
+        getChildren().addListener((Observable obs) -> dispose());
```
Comment (TomasMikula) [inline src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java]: This listener needs to be added _after_ the child list has been populated.

**Comment by JordanMartinez** [comment]
I think I found another issue....

File: src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java (hunk 1 above)
Comment (JordanMartinez) [inline src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java]: Text is still not appearing even after changing that.

Comment (JordanMartinez) [inline src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java]: Ah... I think I get what's happening now.
If you look at the mouse, the scroll bar only barely appears. So I guess it is technically working.
rich text demo_003

Comment (JordanMartinez) [inline src/main/java/org/fxmisc/flowless/VirtualizedScrollPane.java]: If using Flowless 0.4.7, that small scrollbar arrow doesn't appear.

**Comment by JordanMartinez** [comment]
I added these checks to see if the content was being disposed of before actually seeing it, but no text was printed:
````java
getChildren().addListener((Observable obs) -> System.out.println("Children has changed: Called before disposing"));
getChildren().addListener((Observable obs) -> dispose());
getChildren().addListener((Observable obs) -> System.out.println("Children has changed: Called after disposing"));
````

**Comment by JordanMartinez** [comment]
I added a check in `layoutChildren()`:
````java
double w = layoutWidth - vbarWidth;
double h = layoutHeight - hbarHeight;

System.out.println("Content is being resized to " + w + ", " + h);
content.resize(w, h);
````
It kept printing: `Content is being resized to -17.0, -17.0`

**Comment by JordanMartinez** [comment]
Adding an additional check in `layoutChildren()` showed that LayoutBounds width/height was 0/0:
````java
double layoutWidth = getLayoutBounds().getWidth();
double layoutHeight = getLayoutBounds().getHeight();
System.out.println("Layout Bounds is: Width [" + layoutWidth + "] | Height [" + layoutHeight + "]");
````
Prints: `Layout Bounds is: Width [0.0] | Height [0.0]`

**Comment by JordanMartinez** [comment]
I'm pretty sure it's related to the following code. The original VirtualFlow contains this code which is not in the current VirtualizedScrollPane:
````java
getChildren().add(navigator);
clipProperty().bind(Val.map(
    layoutBoundsProperty(),
    b -> new Rectangle(b.getWidth(), b.getHeight())));
````

There's also no way for VirtualizedScrollPane to know about `Navigator` because Virtualized doesn't have such a method.

**Comment by TomasMikula** [comment]
That code was in `VirtualFlowContent` and as far as I can see it _is_ now in `VirtualFlow`.

**Comment by JordanMartinez** [comment]
Any ideas then? Even if I use write `double layoutWidth = content.getLayoutBounds().getWidth()`, and the same for `layoutHeight`, it still doesn't display the text.

**Comment by JordanMartinez** [comment]
That doesn't work either. It makes the scroll bar flicker.

**Comment by JordanMartinez** [comment]
If I add only `VirtualFlow` to children and `layoutChildren()` only has the following code, the text still doesn't display:
````java
double layoutWidth = getLayoutBounds().getWidth();
double layoutHeight = getLayoutBounds().getHeight();
content.resize(layoutWidth, layoutHeight);
````

**Comment by JordanMartinez** [comment]
Figured out the issue. It has nothing to do with VirtualFlow. Sorry for the false alarm.

StyledTextAreaView's layoutChildren() method wasn't updated as well:
````java
@Override
protected void layoutChildren() {
    virtualFlow.resize(getWidth(), getHeight()); // this line needed to be changed to..
    vsPane.resize(getWidth(), getHeight());  // this line. Once changed, the issue is resolved.

    // rest of the method
}
````

**Comment by JordanMartinez** [comment]
K. I'm satisfied with the PR.

**Pull request merged**